### PR TITLE
Refactor to Blade Icons

### PR DESCRIPTION
**Pull request by driesvints**
I'm very happy to say that Adam was so kind to let me take over the library. I recently created my own [Blade Icons](https://github.com/driesvints/blade-icons) package without realizing that Adam already created this wonderful package! After thinking it over I thought it would be perfect if we merged the two packages into one. So that's great news for everyone using this library as I'll be actively maintaining it and helping out with issues etc.

There's some pointers on this so let's go through them. First of all, this PR is a complete rewrite of the package. If I'm going to maintain it and use it as a base for future packages/ideas I'd like to be familiar and comfortable with the library so a rewrite makes sense. You can check out the changes in this pr, [the new README](https://github.com/adamwathan/blade-svg/tree/next-version) on its own branch as well as [the upgrade guide](https://github.com/adamwathan/blade-svg/blob/next-version/UPGRADE.md).

### New Requirements

There's some new minimum version requirements like PHP 7.2 and Laravel 7.14 (I need this version specifically for the Blade components). It doesn't makes sense to me to maintain anything for a PHP version that isn't getting active support anymore.

### New Features

There's new stuff as well! The package now supports multiple icon sets which makes it very easy for third party packages to be created for it. When the refactor has been merged and released I'll release another "Heroicons" package alongside it which you can optionally pull in.

Besides that you can now reference icons as Blade components. This is a very clean syntax and I personally much prefer this over a directive.

### Breaking Changes

There's quite a few breaking changes unfortunately but making these changes now before a full 1.0 release is done will allow me to work with a syntax that I'm comfortable with. Please refer to  [the upgrade guide](https://github.com/adamwathan/blade-svg/blob/next-version/UPGRADE.md) for the most notable ones.

### Package Renaming

One last note I want to make is that this rewrite is part of a bigger plan that I have. Before this PR is merged I'll transfer the repo to a new organisation and rename its package name. I'll give more info as soon as we get to that.

## I need your help

What I need most now is feedback and people testing the library, especially existing users. I'd much appreciate it! Try out this PR by installing it through Composer:

```bash
composer require nothingworks/blade-svg:dev-next-version
```

There's two things I'm currently considering before shipping the next 0.4.0 release:

- ~Should we remove the sprite sheets functionality? It doesn't seems to be like something many people use and would clean up the internals a lot.~ (Sprite Sheets functionality was removed)
- At the moment blade components icons are loaded with loading the file list on every request. It's probably best that we cache this somehow between requests.

I'd much appreciate feedback so let me know what you think!

### CODE REVIEW

**Comment by avvertix** [comment]
Hi @driesvints that's a wonderful new life for this library. Thanks for accepting comments on the topic.

New features like multiple iconset will make a valuable addition especially for defining also illustrations and not only icons. I had a requirement like that when creating an application and I almost created blade components (using the old syntax) to solve it for each illustrations as I was already pulling in an iconset with blade-svg.

> There's some new minimum version requirements like PHP 7.2 and Laravel 7.6 

The new Blade components are definitely a good point to break compatibility, but old Blade directives are still valuable for who uses Laravel 6.x (which if I'm correct is a LTS). I personally think that supporting an LTS version is an option to consider, or at least create an intermediate version that supports it.

> At the moment blade components icons are loaded with loading the file list on every request. It's probably best that we cache this somehow between requests.

I think this was also the approach @adamwathan took. Indeed you have to read the file somewhere unless you cache the entire view in html. Maybe a mad idea could be inlining the SVG in the view cache already generated by Blade instead of the method to read the file. I honestly have no idea how this could be done, but using file cache or redis or the database would not resolve the problem, as it will just read a different "file".

> You can check out the changes in this pr, the new README on its own branch as well as the upgrade guide.

I like the approach for getting icons from different sets by prefixing with the set name.

What if I would like to deliver my iconset package via composer?

Let me explain, I created a package [avvertix/materialicons-laravel-bridge](https://github.com/avvertix/materialicons-laravel-bridge) which was using @adamwathan blade-svg under the hood to deliver the MaterialIcons iconset. Maybe not the most elegant way, but it worked out. Considering your will to rewrite and expand with more iconsets I would love to hear your opinion on additional packages that delivers iconset extensions. Something like :

```
composer require blade-icons-awesome-set
```

```php
'sets' => [
    'awesome-set' => [
        'path' => \AwesomeSet\Providers\Icons::class,
    ],
],
```

and `\AwesomeSet\Providers\Icons::class` is an invokable class returning the path on disk.

Thanks again @driesvints to be open to accept comments.

**Comment by driesvints** [comment]
Hey @avvertix, thanks a bunch for the very thorough feedback!

> I personally think that supporting an LTS version is an option to consider, or at least create an intermediate version that supports it.

Unfortunately we can't do that. I'm making use of a feature in Laravel 7.6 for the Blade components so that's going to have to be the very minimum I'm afraid.

> Indeed you have to read the file somewhere unless you cache the entire view in html.

I already took over the same in-memory caching approach that adam dit for individual svg contents but I'm more talking about the new file-listing that's done when the Blade components are registeren. Maybe we need to create like a manifest. But I'm not sure if it's a big enough problem to worry about.

> Maybe a mad idea could be inlining the SVG in the view cache already generated by Blade instead of the method to read the file.

Hmm, yeah I'm also not sure how that'd be tackled.

> What if I would like to deliver my iconset package via composer?

I'll write a dedicated section on this in the readme soon! After I got some initial feedback here I'm gonna build the heroicons packages. It's actually really easy as all you need to do is to boot the icon set in your service provider.

**Comment by minthemiddle** [comment]
> Should we remove the sprite sheets functionality? It doesn't seems to be like something many people use and would clean up the internals a lot.

Totally, never used it myself nor in my teams.

**Comment by driesvints** [comment]
@minthemiddle thanks. As it stands and from all the feedback I've gotten so far I'm probably going to remove the sprite sheets support this weekend.

**Comment by mvdnbrk** [comment]
I have used this package in the past. Never used the sprite sheets as well.
Good to hear this package gets some fresh air 👍

Came back here for a new project I am working on. Will pull in the PR to try it out.
Thanks Dries!

**Comment by driesvints** [comment]
Heya, some updates:

- I've decided to remove sprite sheets functionality as I get the feeling that they're not much used at all.
- Added a section about how to build a third party package for your icons.

I feel like we're pretty close to wrapping things up now. The most thing that's needed atm is people testing it.

**Comment by njoguamos** [comment]
I have been using this package for quite sometimes. I have the `dev-next-version` a spin and here are my quick thoughts.


1.  Though renaming `svg_image()` helper to `svg()` is a breaking change, I support it, as a small price to pay 😀.
2. I totally support removal of sprite sheets, I personally never used theme before.
3. The new blade components comes in handy. However I am having problem loading icons with it. 
```blade
{{-- This is working --}}
@svg('search','h-4 w-4'')

{{-- This is working too --}}
{{ svg('search','h-4 w-4') }}

{{-- This is NOT working --}}
<x-icon-search class="h-5 w-5"/>
```
 Am getting an error `Unable to locate a class or view for component [icon-search]`. I will I investigate my codebase and open an issue if necessary and share details.

4. Finally, am quite interested on how caching icons can be achieved. Am however happy to use the package to render svg and focus on optimising performance on other key areas such as database queries.


@driesvints keep pumping life to this package.

**Comment by bambamboole** [comment]
@driesvints I tried to use the package inside another laravel package and it worked out of the box with a custom set.
Can't wait fo the new release. Good work!

File: README.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+If you're interested in building your own third party package to integrate an icon set, it's pretty easy to do so. Make sure to load your SVGs from the `boot` method of your package's service provider. Provide the path to your SVGs and provide your own unique set name and component prefix:
+
+```php
+use BladeUI\Icons\Factory;
+
+public function boot(): void
+{
+    $this->app->make(Factory::class)->add('heroicons', [
+        'path' => __DIR__ . '/../resources/svg',
+        'component-prefix' => 'heroicon',
```
Comment (avvertix) [inline README.md]: @driesvints I was trying to follow the package creation and I guess it should be `prefix` and not `component-prefix` the configuration key to use

```suggestion
        'prefix' => 'heroicon',
```

Comment (driesvints) [reply README.md]: Thanks! Fixed.

**Comment by driesvints** [comment]
I've made some updates today that generalises the usages of prefixes across the directives and components.

I also discovered a bug with Laravel which makes it impossible to actually use the components right now. I'll try to dig into that to see if I can solve it in the framework this week.

**Comment by meduzen** [comment]
_(Finally taking time to send feedback.)_

A few years ago I dropped `blade-svg`. Here’s how I dealt with SVG since then.

## Intro

I put SVGs in Blade files so that I can `@include('svg.my-icon')`, sometimes passing more data to the SVG Blade view, allowing high customization.

Eventually I planned to make a directive alias (`@svg('my-icon')`), or a slotted `@component` (now `<x-something>`) for when I need custom DOM.

## Not using spritesheets

Example 1 (most current use): similarly to blade-svg, it can receive custom CSS classes, width and height.

```blade
{{-- svg/basecamp-logo.blade.php --}}
<svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 36.83 32.08"
    width="{{ $width ?? 27 }}" height="{{ $height ?? 24 }}"
    fill="none"
    stroke="currentColor"
    stroke-linecap="round" stroke-linejoin="round" stroke-width="3.41"
    class="{{ $class ?? '' }} basecamp-svg"
>
    <path d="m1.71 17.73c2.61-7.4 7.61-16 17.7-16s15.08 13.27 15.71 22.27c-2.65 4.6-9 6.33-15.71 6.33a18.87 18.87 0 0 1 -14.41-6.33s3.16-8.53 6.11-8.55c2.17 0 4 4 5.73 4s5.33-7.68 5.33-7.68"/>
</svg>
```

```blade
<a href="{{ route('basecampConnect') }}">
    Login with Basecamp
    @include('svg.basecamp', ['class' => 'login__basecamp-svg'])
</a>
```

Example 2: optional DOM element (here: date on a calendar icon)

```blade
{{-- svg/calendar.blade.php --}}
<svg>
    <path d="M6 5V1v4zm8 0V1v4zM5 9h10H5zM3 19h14a2 2 0 002-2V5a2 2 0 00-2-2H3a2 2 0 00-2 2v12c0 1.1.9 2 2 2z"/>
    @isset ($day)
        <text>{{ today()->day }}</text>
    @endisset
</svg>
```

```blade
<a class="calendar__control | btn" id="today-btn" href="#today" title="Scroll to today">
    @include('svg.calendar', ['day' => today()->day])
</a>
```

image

(Yes, it’s the [Heroicons one](https://github.com/refactoringui/heroicons/blob/master/src/outline-md/md-calendar.svg), modified.)

## Using spritesheets

_(I don’t use spritesheets on most projects, but sometimes they make the DOM way lighter.)_

I `@include('spritesheet.blade.php')` before closing the `</body>` of the layout template:

```blade
{{-- spritesheet.blade.php --}}
<div class="none">
    @include('svg.calendar')
    @include('svg.chevron')
    @include('svg.envelope')
    @include('svg.lock-esporg')
    @include('svg.map-marker')
    @include('svg.phone')
</div>
```

Each of the `@include('svg.…')` brings a full SVG:

```blade
{{-- svg/calendar.blade.php --}}
<svg id="calendar-path" other-attributes…>
    <path…>
</svg>
```

For each of the _complete_ SVGs, there’s another file with `<use>`, ready to be included by other templates:

```blade
{{-- icon/calendar.blade.php --}}
<svg attributes…>
    <use xlink:href="#calendar-path"/>
</svg>
```

```blade
{{-- then in page.blade.php --}}
@include('icon.calendar', ['class' => 'booking__icon'])
```

## In short

This way of working allows high customisation. As mentioned earlier in this message, improvements I see are:
- aliasing the `@include('path/to/svg.…)`
- using Blade components, maybe slotted.
- (if not a perf killer) a `@spritesheet('path/to/a/lot/of/svg')` directive to generate a spritesheet from a directory of SVGs.

## Regarding [next-version](https://github.com/adamwathan/blade-svg/tree/7272477866bdfd8d580e9c6a0a4cc94588c9c921)

I think it’s a solid refresh with good flexibility for a lot of uses cases. Automatic declaration of the Blade components for each set is really the feature that could be missing from my workflow.

What is missing for me to use it is the customization on the DOM part (and now, the spritesheet :D), that I need for most projects.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        'default' => [
+            'prefix' => 'icon',
+        ],
+    ],
+];
+```
+
+Defining a prefix for every set is required and every prefix should be unique.
+
+When referencing icons with the [Blade directive](#directives) or [helper](#helper) you can omit the prefix to reference icons from the `default` set.
```
Comment (njoguamos) [inline README.md]: I think it should be worth noting that icons named in kebab-case MUST include the prefix.

For example, for an icon `/resources/svg/arrow-left-circle.svg`

```blade
{{-- This wont work. An error 'Svg by name "right" from set "default" not found' will be thrown--}}
@svg('arrow-left-circle','h-5 text-gray-700')

{{-- This will work --}}
@svg('icon-arrow-left-circle','h-5 text-gray-700')
```

The alternative is to ensure that all icons ARE NOT kebab-cased, instead use camelCase or PascalCase or snake_case. This maybe challenging considering most icon set use kebab-case naming.

Comment (driesvints) [reply README.md]: Ow! You've actually discovered a bug. I fixed this. Default icons don't need the prefix anymore. Thanks for testing!

**Comment by driesvints** [comment]
@meduzen I'm actually not entirely opposed to bringing back sprite sheet support. After all it was already done. But before I do that could you please walk me through why you need it in this day and age?

I think the idea about slotted components is a good one and one I'm willing to explore in a future version.

PS.: I've finally fixed the component name issue in Laravel so Blade components should work again soon. ~I'll try to tackle this in the weekend.~ Need to wait until a new release gets tagged on Tuesday.

File: README.md (hunk 2 above)
Comment (njoguamos) [reply README.md]: You are welcome. Thanks for fixing.

**Comment by agentphoenix** [comment]
Loving the changes to the package, Dries!

I pulled the branch down and played around with it today and noticed one bug. If you specify a default class in the config file, it will only be applied if you set a class when you use the icon. If you use a blank string (which is the default), it never gets applied. I like using a default class to set some basic styles for icons (height, width, etc.) so that jumped out at me right away.

Not a bug, but would be nice to have, is a way to specify additional classes that are specific to a certain set. For example, one of my sets is a stroke set and the other is a filled set. I'd love to be able to specify `stroke-current` for the stroke set and `fill-current` for the filled set right from the config file. Not a dealbreaker, but definitely would be nice to have.

**Comment by driesvints** [comment]
@agentphoenix thanks!

> I pulled the branch down and played around with it today and noticed one bug. If you specify a default class in the config file, it will only be applied if you set a class when you use the icon. If you use a blank string (which is the default), it never gets applied. I like using a default class to set some basic styles for icons (height, width, etc.) so that jumped out at me right away.

Hmm, not sure I understand. Can you provide a code sample? 

> Not a bug, but would be nice to have, is a way to specify additional classes that are specific to a certain set. For example, one of my sets is a stroke set and the other is a filled set. I'd love to be able to specify stroke-current for the stroke set and fill-current for the filled set right from the config file. Not a dealbreaker, but definitely would be nice to have.

Definitely something we can consider. Can you share a code sample of this as well how'd you exactly use this?

**Comment by driesvints** [comment]
I've just pushed a commit that fixes the component rendering. Components should now properly be handled. Unfortunately I had to bump the minimum requirements to the just-released Laravel v7.10.0 release. 

Can anyone try out the components and let me know how it goes?

**Comment by driesvints** [comment]
@meduzen I'd still be keen to bringing back sprite sheets but can you let me know about my question here? https://github.com/adamwathan/blade-svg/pull/50#issuecomment-621402470

**Comment by agentphoenix** [comment]
@driesvints [This](https://github.com/adamwathan/blade-svg/blob/e17192c6244b0fd9ed8b8241e7d4a1328478c498/src/Factory.php#L140) is the line that causes the issue as far as I could tell. Here's an example of what I was seeing:

```php
// Will not apply the default class from the config file
@svg('house')

// Will apply the default class from the config file
@svg('house', 'h-5 w-5')
```
For the set classes, I figured it'd be simplest to do something like this:

```php
'sets' => [
    'heroicons-solid' => [
      'path' => '...',
      'prefix' => 'heroicons',
      'class' => 'fill-current',
    ],
    'heroicons-solid' => [
      'path' => '...',
      'prefix' => 'heroicons-outline',
      'class' => 'stroke-current',
    ],
]
```
From there, the package could merge the default class and the set class. (It also wouldn't be a huge deal if the set class took priority and I needed to set the default class multiple times.) Does that make sense?

**Comment by njoguamos** [comment]
> I've just pushed a commit that fixes the component rendering. Components should now properly be handled. Unfortunately I had to bump the minimum requirements to the just-released Laravel v7.10.0 release.
> 
> Can anyone try out the components and let me know how it goes?

Managed to pull the latest commit. Unfortunately, the blade component is not working.

```blade
{{-- This is working --}}
@svg('filter','h-5 fill-current')

{{-- This is working --}}
{{ svg('filter','h-5 fill-current') }}

{{-- This is failing --}}
 <x-icon-filter class="h-5 fill-current"/>
```
```bash
Unable to locate a class or view for component [icon-filter].
```

Am Using

- Laravel version 7.10.1
- Default icon set

**Comment by njoguamos** [comment]
I have checked into @agentphoenix issue and look like we have un-expected behaviour.

...blade-icons.php
```php
return [
//....
'class' => 'h-4 text-gray-500',
]
```
This syntax 
```blade
@svg('filter','h-10 text-success-700')
{{ svg('filter','h-10 text-success-700') }}
```
Is rendering
```html
<svg class="h-4 text-gray-500 h-10 text-success-700" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">....</svg>
```
While this
```blade
 @svg('filter')
{{ svg('filter') }}
```
Is rendering without the default class
```html
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512">....</svg>
```

I also think @agentphoenix idea of **having default class per icon set** is more flexible than having a global default class.

**Comment by driesvints** [comment]
@agentphoenix @njoguamos okay, I've managed to fix the default classes and I've also added the ability to add default classes per-set. Make sure to re-publish your config files when testing.

> Managed to pull the latest commit. Unfortunately, the blade component is not working.

Hmm, that's strange. Your code samples are working for me. Did you clear your views? You're sure your icons are place in the correct directory and named correctly? Can you republish your config file? Did you uncomment the options inside?

**Comment by njoguamos** [comment]
@driesvints I will give it another try and let you know.

**Comment by agentphoenix** [comment]
@driesvints Works like a charm! Thanks for the quick turnaround on this.

**Comment by driesvints** [comment]
Unless there's any more feedback I'm probably going to be moving the package to its new location and doing the 0.4.0 release somewhere next week.

**Comment by njoguamos** [comment]
@driesvints I re-published the config file, cleared view and pulled the latest commit

> Hmm, that's strange. Your code samples are working for me. Did you clear your views? You're sure your icons are place in the correct directory and named correctly? Can you republish your config file? 

> Did you uncomment the options inside?


My config
```php
return [
    'sets' => [
        'default' => [
            'path' => 'resources/icons',
            'prefix' => 'icon',
            'class' => 'h-5 text-gray-130 fill-current',
        ],
    ],

    'class' => '',
];
```

Everything is working as expected except the blade component. The default set class `h-5 text-gray-130 fill-current` is not being applied on blade component.

```blade
// This applies
@svg('icon-globe')

// This applies
{{ svg('icon-globe') }}

// This renders svg but doesn't apply 
<x-icon-globe></x-icon-globe>

// This applies
<x-icon-globe class="h-5 fill-current text-gray-130"></x-icon-globe>
```

And thank for making the blade components work. Effort appreciated.

**Comment by agentphoenix** [comment]
I ran into a problem tonight running my test suite in Github Actions with this package. From the exceptions thrown, it seems that the package can't properly parse out the set from the icon name. Since I don't use a set named `default`, my tests blow up saying it can't find the icon `fluent-eye-off` in the set `default`. Instead, it should be looking for the icon `eye-off` in the set `fluent`.

Everything works fine locally and on the test server I'm running, but Github Actions throws an exception the first time I do `$this->get` from my feature test. Any ideas why that might be happening? Any ideas for how to best debug this?

Screen Shot 2020-06-16 at 12 02 58 AM

**Comment by driesvints** [comment]
I'll take a look at these issues tomorrow and will try to solve them. Thanks a lot for testing btw!

If everything goes well I might also finally move the repo and release the new version tomorrow.

**Comment by driesvints** [comment]
Hey @njoguamos. Thank you so much for helping with testing. I pushed a commit that will fix this. You might have to re-run composer install and clear your views to test again.

@agentphoenix hmm I can't reproduce that. I've added two tests to properly test the scenario without a default set present: https://github.com/adamwathan/blade-svg/pull/50/commits/8d531f4870970e9cddc6a9be5384b4e4fdc9c487

Can you maybe share the code to reproduce this?

**Comment by agentphoenix** [comment]
@driesvints Unfortunately I can't reproduce the issue in my local tests, on my local server, or on my staging server. When I run the test suite through Github Actions though, it blows up. Here's a link to the Github Action result where you see the exception.

https://github.com/anodyne/nova3/runs/780876889?check_suite_focus=true

I'm just more curious about ways I can debug what's actually happening with GA so that I can figure out how to reproduce this locally and see if it's something with my app or with the package.

**Comment by driesvints** [comment]
@agentphoenix I see what's wrong. You're not using your set prefix when referencing the icons: https://github.com/anodyne/nova3/blob/dev/nova/resources/views/components/input/password.blade.php#L20

I think you missed this part in the readme:

> When referencing icons with the Blade directive or helper you can omit the prefix to reference icons from the default set.

I'll try to make that even more clear by adding a sentence that says the prefix is mandatory for other sets.

**Comment by agentphoenix** [comment]
@driesvints So I have a custom directive `@icon` that I use since users can swap the icon set they use in the app. Behind the scenes, it takes the name passed into the directive, does a lookup on an icon set map, and then prepends the set prefix and calls the Svg class. The same error gets thrown if I do `@svg('fluent-eye-off')` as well.

**Comment by driesvints** [comment]
@agentphoenix we can't offer support for custom directives sorry.

> The same error gets thrown if I do @svg('fluent-eye-off') as well.

Hmm that's odd because this test verifies that it should work: https://github.com/adamwathan/blade-svg/pull/50/commits/8d531f4870970e9cddc6a9be5384b4e4fdc9c487#diff-b0e3204a285e6556082196cf2a8dc305R125

**Comment by driesvints** [comment]
Hey everyone. We've finally merged this and moved the package to its new home. I'll be leaving this branch on this repo for a few weeks but it's best that you move to the stable 0.4.0 tag as soon as possible.

Thanks a lot for helping with testing!

**Comment by agentphoenix** [comment]
@driesvints Yeah no, I wouldn't expect support for the custom directive. It's just weird because everything is being built up properly and even with the `@svg` directive the tests are failing, so I'm just at a loss for how to debug it. No worries though, I'll keep hammering on it. Thanks for all your work on this!

**Comment by njoguamos** [comment]
@driesvints  thank you for taking your time to refine the package. 

@agentphoenix I understand what you are facing. I have something similar. Everything works on local server but failing on github action. Keep working. If I have sometimes, I will have a look at your issue.

**Comment by agentphoenix** [comment]
@driesvints @njoguamos I ended up fixing my issue with Github Actions, though I'm not sure _why_ it fixed it, so maybe one of you knows.

I ended up dumping out the registered sets from the BladeUI icons factory class in my tests. Locally, everything looked as expected. In Github Actions though, it was an empty array. So clearly something was going on where Github Actions was running my tests before that service provider could be called. On a whim, I decided to turn off Laravel's auto-discovery and manually registered the BladeUI icons service provider myself. Everything worked after that.

So the question now is... why? Not sure if there's something about Laravel's auto-discovery and when service providers get called, but this is the second time that removing something from auto-discovery fixed my tests in Github Actions. Welcome to hear thoughts on that if you have any, but either way, thought I'd let you both know that I fixed it and what the fix was.

**Comment by driesvints** [comment]
@agentphoenix hmm no idea really. I never experienced that myself. Good that you got it fixed though.
